scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/TreeOrderedBuf.scala

```diff
@@ -0,0 +1,309 @@
+/*
+ Copyright 2014 Twitter, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordered_serialization
+
+import scala.reflect.macros.Context
+import scala.language.experimental.macros
+import java.io.InputStream
+
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.OrderedSerialization
+object CommonCompareBinary {
+  import com.twitter.scalding.serialization.JavaStreamEnrichments._
+
+  // If the lengths are equal and greater than this number
+  // we will compare on all the containing bytes
+  val minSizeForFulBinaryCompare = 24
+
+  final def compareBinaryPrelude(inputStreamA: InputStream,
+    lenA: Int,
+    inputStreamB: InputStream,
+    lenB: Int)(innerCmp: (InputStream, InputStream) => Int) = {
+    try {
+      // First up validate the lengths passed make sense
+      require(lenA >= 0, "Length was " + lenA + "which is < 0, invalid")
+      require(lenB >= 0, "Length was " + lenB + "which is < 0, invalid")
+
+      val earlyEqual: Boolean = if (lenA > minSizeForFulBinaryCompare &&
+        (lenA == lenB) &&
+        inputStreamA.markSupported &&
+        inputStreamB.markSupported) {
+        inputStreamA.mark(lenA)
+        inputStreamB.mark(lenB)
+
+        var pos = 0
+        var isSame = true
+
+        while (pos < lenA && isSame == true) {
+          inputStreamA.readByte == inputStreamB.readByte
+          pos = pos + 1
+        }
+
+        if (isSame) {
+          isSame
+        } else {
+          // rewind if they don't match for doing the full compare
+          inputStreamA.reset()
+          inputStreamA.reset()
+          isSame
+        }
+      } else false
+
+      val r = if (earlyEqual) {
+        0
+      } else {
+        val bufferedStreamA = _root_.com.twitter.scalding.serialization.PositionInputStream(inputStreamA)
```

Comment:
Yeah its all been pulled out of the macro. It probably should go somewhere else in serialization land since its nothing specific to do with this